test/integration/integration_test.cc

```diff
@@ -483,6 +483,37 @@ TEST_P(IntegrationTest, TestSmuggling) {
   }
 }
 
+TEST_P(IntegrationTest, TestPipelinedResponses) {
+  initialize();
+  auto tcp_client = makeTcpConnection(lookupPort("http"));
+
+  ASSERT_TRUE(tcp_client->write(
+      "POST /test/long/url HTTP/1.1\r\nHost: host\r\ntransfer-encoding: chunked\r\n\r\n"));
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  std::string data;
+  ASSERT_TRUE(fake_upstream_connection->waitForData(
+      FakeRawConnection::waitForInexactMatch("\r\n\r\n"), &data));
+  ASSERT_THAT(data, HasSubstr("POST"));
+
+  ASSERT_TRUE(fake_upstream_connection->write(
+      "HTTP/1.1 200 OK\r\ntransfer-encoding: chunked\r\n\r\n0\r\n\r\n"
+      "HTTP/1.1 200 OK\r\ntransfer-encoding: chunked\r\n\r\n0\r\n\r\n"
+      "HTTP/1.1 200 OK\r\ntransfer-encoding: chunked\r\n\r\n0\r\n\r\n"));
+
+  tcp_client->waitForData("\r\n\r\n", false);
+  std::string response = tcp_client->data();
+
+  EXPECT_THAT(response, HasSubstr("HTTP/1.1 200 OK\r\n"));
+  EXPECT_THAT(response, HasSubstr("transfer-encoding: chunked\r\n"));
+  EXPECT_THAT(response, EndsWith("\r\n\r\n"));
+
+  ASSERT_TRUE(fake_upstream_connection->close());
```

Comment:
Can we verify that there's a single reply to the request instead of 3?

Reply:
brainstorming. right now added and tested a stat increment. 
codec_client_test wouldn't reveal extra information.

it's tough because waitfor will wait for data, but doesn't mean could always be the end in this case. to some degree, we hit all sorts of assert failures if we try to even process > 1.

Reply:
I think that the tough part is that we can't just do a strict string compare on the response because the proxy adds some extra headers that are not easily predictable.  I think that the EndsWith check you added gets us most of the way.

Reply:
yes, i can't do a strict match because of time changing things like dates in the added headers. I could make part of the body be response number specific (e.g. "first response").

I think my issue that you could still have the pipelined responses later on is that I need to add more to this test that asserts that a follow-up request will not use the second pipelined response, but rather the next response.